packages/migrate/migrations/svelte-4/migrate.js

```diff
@@ -0,0 +1,200 @@
+import fs from 'node:fs';
+import { Project, ts, Node } from 'ts-morph';
+
+/** @param {string} file_path */
+export function update_svelte_file(file_path) {
+	const content = fs.readFileSync(file_path, 'utf-8');
+	const updated = content.replace(
+		/<script([^]*?)>([^]+?)<\/script>(\n*)/g,
+		(_match, attrs, contents, whitespace) => {
+			return `<script${attrs}>${transform_code(contents)}</script>${whitespace}`;
+		}
+	);
+	fs.writeFileSync(file_path, transform_svelte_code(updated), 'utf-8');
+}
+
+/** @param {string} file_path */
+export function update_js_file(file_path) {
+	const content = fs.readFileSync(file_path, 'utf-8');
+	const updated = transform_code(content);
+	fs.writeFileSync(file_path, updated, 'utf-8');
+}
+
+/** @param {string} code */
+export function transform_code(code) {
+	const project = new Project({ useInMemoryFileSystem: true });
+	const source = project.createSourceFile('svelte.ts', code);
+	update_imports(source);
+	update_typeof_svelte_component(source);
+	update_action_types(source);
+	update_action_return_types(source);
+	return source.getFullText();
+}
+
+/** @param {string} code */
+export function transform_svelte_code(code) {
+	return update_transitions(update_svelte_options(code));
+}
+
+/**
+ * <svelte:options tag=".." /> -> <svelte:options customElement=".." />
+ * @param {string} code
+ */
+function update_svelte_options(code) {
+	return code.replace(/<svelte:options([^]*?)\stag=([^]*?)\/?>/, (match) => {
+		return match.replace('tag=', 'customElement=');
+	});
+}
+
+/**
+ * transition/in/out:x -> transition/in/out:x|global
+ * @param {string} code
+ */
+function update_transitions(code) {
+	return code.replace(/(\s)(transition:|in:|out:)(\w+)(?=[\s>=])/g, '$1$2$3|global');
```

Comment:
per dominik's suggestion on Discord: this might be tedious to undo and most people probably won't want `|global` even if that's the same behavior as svelte 3. maybe either remove it or make it an option